README.md
# Introduction
This repository contains an Erlang module to deal with base64 and
base64url encoding.

# Build
You can build the library with:

    make build

# Test
You can execute the test suite with:

    make dialyzer test

You can generate test coverage with:

    make cover

# Documentation
A handbook is available [in the `doc` directory](doc/handbook.md).

# Contact
If you find a bug or have any question, feel free to open a Github
issue.

Please not that we do not currently review or accept any contribution.

# License
Released under the ISC license.

Copyright (c) 2022 Bryan Frimin <bryan@frimin.fr>.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


CHANGELOG.md
# Introduction

All notable changes to this project will be documented in this file.

The format is based on [Keep a
Changelog](https://keepachangelog.com/en/1.0.0/), and this project
adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 1.1.1 (2022-07-26)

No change.

## 1.1.0 (2021-07-29)

### Added

- Export `decode_error_reason/0` type.

## 1.0.0 (2021-06-14)

No change.

## 0.1.0 (2021-02-04)

### Added

- Add `b64:encode/1` and `b64:encode/2` functions.
- Add `b64:decode/1` and `b64:decode/2` functions.
- Add `b64url:encode/1` and `b64url:encode/2` functions.
- Add `b64url:decode/1` and `b64url:decode/2` functions.



test/b64_tests.erl
%% Copyright (c) 2022 Bryan Frimin <bryan@frimin.fr>.
%% Copyright (c) 2021 Exograd SAS.
%% Copyright (c) 2021 Bryan Frimin <bryan@frimin.fr>.
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
%% SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
%% IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(b64_tests).

-include_lib("eunit/include/eunit.hrl").

encode_without_padding_test_() ->
  [?_assertEqual(<<>>,
                 b64:encode(<<>>, [nopad])),
   ?_assertEqual(<<"Zg">>,
                 b64:encode(<<"f">>, [nopad])),
   ?_assertEqual(<<"Zm8">>,
                 b64:encode(<<"fo">>, [nopad])),
   ?_assertEqual(<<"Zm9v">>,
                 b64:encode(<<"foo">>, [nopad])),
   ?_assertEqual(<<"Zm9vYg">>,
                 b64:encode(<<"foob">>, [nopad])),
   ?_assertEqual(<<"Zm9vYmE">>,
                 b64:encode(<<"fooba">>, [nopad])),
   ?_assertEqual(<<"Zm9vYmFy">>,
                 b64:encode(<<"foobar">>, [nopad]))].

encode_with_padding_test_() ->
  [?_assertEqual(<<>>,
                 b64:encode(<<>>)),
   ?_assertEqual(<<"Zg==">>,
                 b64:encode(<<"f">>)),
   ?_assertEqual(<<"Zm8=">>,
                 b64:encode(<<"fo">>)),
   ?_assertEqual(<<"Zm9v">>,
                 b64:encode(<<"foo">>)),
   ?_assertEqual(<<"Zm9vYg==">>,
                 b64:encode(<<"foob">>)),
   ?_assertEqual(<<"Zm9vYmE=">>,
                 b64:encode(<<"fooba">>)),
   ?_assertEqual(<<"Zm9vYmFy">>,
                 b64:encode(<<"foobar">>))].

decode_without_padding_test_() ->
  [?_assertEqual({ok, <<>>},
                 b64:decode(<<>>, [nopad])),
   ?_assertEqual({ok, <<"f">>},
                 b64:decode(<<"Zg">>, [nopad])),
   ?_assertEqual({ok, <<"fo">>},
                 b64:decode(<<"Zm8">>, [nopad])),
   ?_assertEqual({ok, <<"foo">>},
                 b64:decode(<<"Zm9v">>, [nopad])),
   ?_assertEqual({ok, <<"foob">>},
                 b64:decode(<<"Zm9vYg">>, [nopad])),
   ?_assertEqual({ok, <<"fooba">>},
                 b64:decode(<<"Zm9vYmE">>, [nopad])),
   ?_assertEqual({ok, <<"foobar">>},
                 b64:decode(<<"Zm9vYmFy">>, [nopad])),
   ?_assertEqual({error, {invalid_data, <<"!">>}},
                 b64:decode(<<"!">>, [nopad])),
   ?_assertEqual({error, {invalid_base64_char, 33}},
                 b64:decode(<<"a!">>, [nopad])),
   ?_assertEqual({error, invalid_encoding},
                 b64:decode(<<"Zg==">>, [nopad]))].

decode_with_padding_test_() ->
  [?_assertEqual({ok, <<>>},
                 b64:decode(<<>>)),
   ?_assertEqual({ok, <<"f">>},
                 b64:decode(<<"Zg==">>)),
   ?_assertEqual({ok, <<"fo">>},
                 b64:decode(<<"Zm8=">>)),
   ?_assertEqual({ok, <<"foo">>},
                 b64:decode(<<"Zm9v">>)),
   ?_assertEqual({ok, <<"foob">>},
                 b64:decode(<<"Zm9vYg==">>)),
   ?_assertEqual({ok, <<"fooba">>},
                 b64:decode(<<"Zm9vYmE=">>)),
   ?_assertEqual({ok, <<"foobar">>},
                 b64:decode(<<"Zm9vYmFy">>)),
   ?_assertEqual({error, {invalid_data, <<"!">>}},
                 b64:decode(<<"!">>)),
   ?_assertEqual({error, {invalid_base64_char, 36}},
                 b64:decode(<<"ZjA$">>))].


test/b64url_tests.erl
%% Copyright (c) 2022 Bryan Frimin <bryan@frimin.fr>.
%% Copyright (c) 2021 Exograd SAS.
%% Copyright (c) 2021 Bryan Frimin <bryan@frimin.fr>.
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
%% SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
%% IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(b64url_tests).

-include_lib("eunit/include/eunit.hrl").

encode_without_padding_test_() ->
  [?_assertEqual(<<>>,
                 b64url:encode(<<>>, [nopad])),
   ?_assertEqual(<<"Zg">>,
                 b64url:encode(<<"f">>, [nopad])),
   ?_assertEqual(<<"Zm8">>,
                 b64url:encode(<<"fo">>, [nopad])),
   ?_assertEqual(<<"Zm9v">>,
                 b64url:encode(<<"foo">>, [nopad])),
   ?_assertEqual(<<"Zm9vYg">>,
                 b64url:encode(<<"foob">>, [nopad])),
   ?_assertEqual(<<"Zm9vYmE">>,
                 b64url:encode(<<"fooba">>, [nopad])),
   ?_assertEqual(<<"Zm9vYmFy">>,
                 b64url:encode(<<"foobar">>, [nopad]))].

encode_with_padding_test_() ->
  [?_assertEqual(<<>>,
                 b64url:encode(<<>>)),
   ?_assertEqual(<<"Zg==">>,
                 b64url:encode(<<"f">>)),
   ?_assertEqual(<<"Zm8=">>,
                 b64url:encode(<<"fo">>)),
   ?_assertEqual(<<"Zm9v">>,
                 b64url:encode(<<"foo">>)),
   ?_assertEqual(<<"Zm9vYg==">>,
                 b64url:encode(<<"foob">>)),
   ?_assertEqual(<<"Zm9vYmE=">>,
                 b64url:encode(<<"fooba">>)),
   ?_assertEqual(<<"Zm9vYmFy">>,
                 b64url:encode(<<"foobar">>))].

decode_without_padding_test_() ->
  [?_assertEqual({ok, <<>>},
                 b64url:decode(<<>>, [nopad])),
   ?_assertEqual({ok, <<"f">>},
                 b64url:decode(<<"Zg">>, [nopad])),
   ?_assertEqual({ok, <<"fo">>},
                 b64url:decode(<<"Zm8">>, [nopad])),
   ?_assertEqual({ok, <<"foo">>},
                 b64url:decode(<<"Zm9v">>, [nopad])),
   ?_assertEqual({ok, <<"foob">>},
                 b64url:decode(<<"Zm9vYg">>, [nopad])),
   ?_assertEqual({ok, <<"fooba">>},
                 b64url:decode(<<"Zm9vYmE">>, [nopad])),
   ?_assertEqual({ok, <<"foobar">>},
                 b64url:decode(<<"Zm9vYmFy">>, [nopad])),
   ?_assertEqual({error, {invalid_data, <<"!">>}},
                 b64url:decode(<<"!">>, [nopad])),
   ?_assertEqual({error, {invalid_base64_char, 33}},
                 b64url:decode(<<"a!">>, [nopad])),
   ?_assertEqual({error, invalid_encoding},
                 b64url:decode(<<"Zg==">>, [nopad]))].

decode_with_padding_test_() ->
  [?_assertEqual({ok, <<>>},
                 b64url:decode(<<>>)),
   ?_assertEqual({ok, <<"f">>},
                 b64url:decode(<<"Zg==">>)),
   ?_assertEqual({ok, <<"fo">>},
                 b64url:decode(<<"Zm8=">>)),
   ?_assertEqual({ok, <<"foo">>},
                 b64url:decode(<<"Zm9v">>)),
   ?_assertEqual({ok, <<"foob">>},
                 b64url:decode(<<"Zm9vYg==">>)),
   ?_assertEqual({ok, <<"fooba">>},
                 b64url:decode(<<"Zm9vYmE=">>)),
   ?_assertEqual({ok, <<"foobar">>},
                 b64url:decode(<<"Zm9vYmFy">>)),
   ?_assertEqual({error, {invalid_data, <<"!">>}},
                 b64url:decode(<<"!">>)),
   ?_assertEqual({error, {invalid_base64_char, 36}},
                 b64url:decode(<<"ZjA$">>))].


src/b64.erl
%% Copyright (c) 2022 Bryan Frimin <bryan@frimin.fr>.
%% Copyright (c) 2021 Exograd SAS.
%% Copyright (c) 2021 Bryan Frimin <bryan@frimin.fr>.
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
%% SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
%% IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(b64).

-export([encode/1, encode/2,
         decode/1, decode/2]).

-export_type([decode_error_reason/0]).

-type decode_error_reason() ::
        invalid_encoding
      | {invalid_data, binary()}
      | {invalid_base64_char, byte()}.

-spec encode(binary()) -> binary().
encode(Bin) when is_binary(Bin) ->
  encode(Bin, []).

-spec encode(binary(), Options) -> binary() when
    Options :: [Option],
    Option :: nopad.
encode(Bin, Options) when is_binary(Bin), is_list(Options) ->
  case proplists:get_bool(nopad, Options) of
    true ->
      encode(nopad, Bin, <<>>);
    false ->
      encode(pad, Bin, <<>>)
  end.

-spec encode(pad | nopad, binary(), binary()) -> binary().
encode(_Mode, <<>>, Acc) ->
  Acc;
encode(Mode, <<A0:6, B0:6, C0:6, D0:6, Rest/binary>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0),
  D = dec_base64_digit(D0),
  encode(Mode, Rest, <<Acc/binary, A, B, C, D>>);
encode(nopad, <<A0:6, B0:2>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0 bsl 4),
  encode(nopad, <<>>, <<Acc/binary, A, B>>);
encode(nopad, <<A0:6, B0:6, C0:4>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0 bsl 2),
  encode(nopad, <<>>, <<Acc/binary, A, B, C>>);
encode(pad, <<A0:6, B0:2>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0 bsl 4),
  encode(pad, <<>>, <<Acc/binary, A, B, $=, $=>>);
encode(pad, <<A0:6, B0:6, C0:4>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0 bsl 2),
  encode(pad, <<>>, <<Acc/binary, A, B, C, $=>>).

-spec dec_base64_digit(0..63) ->
        $A..$Z | $a..$z | $0..$9 | $+ | $/.
dec_base64_digit(Char) when Char =< 25 ->
  Char + $A;
dec_base64_digit(Char) when Char =< 51 ->
  Char + $a - 26;
dec_base64_digit(Char) when Char =< 61 ->
  Char + $0 - 52;
dec_base64_digit(62) ->
  $+;
dec_base64_digit(63) ->
  $/.

-spec decode(binary()) ->
        {ok, binary()} | {error, decode_error_reason()}.
decode(Bin) when is_binary(Bin) ->
  decode(Bin, []).

-spec decode(binary(), Options) ->
        {ok, binary()} | {error, decode_error_reason()}
          when Options :: [Option],
               Option :: nopad.
decode(Bin, Options) when is_binary(Bin), is_list(Options) ->
  try
    case proplists:get_bool(nopad, Options) of
      true ->
        decode(nopad, Bin, <<>>);
      false ->
        decode(pad, Bin, <<>>)
    end
  catch
    throw:{error, Reason} ->
      {error, Reason}
  end.

-spec decode(pad | nopad, binary(), binary()) ->
        {ok, binary()} | {error, term()}.
decode(_Mode, <<>>, Acc) ->
  {ok, Acc};
decode(nopad, <<_A:8, _B:8, _C:8, $=:8>>, _Acc) ->
  {error, invalid_encoding};
decode(nopad, <<_A:8, _B:8, $=:8, $=:8>>, _Acc) ->
  {error, invalid_encoding};
decode(nopad, <<A0:8, B0:8, C0:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0) bsr 2,
  decode(nopad, <<>>, <<Acc/binary, A:6, B:6, C:4>>);
decode(nopad, <<A0:8, B0:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0) bsr 4,
  decode(nopad, <<>>, <<Acc/binary, A:6, B:2>>);
decode(pad, <<A0:8, B0:8, $=:8, $=:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0) bsr 4,
  decode(pad, <<>>, <<Acc/binary, A:6, B:2>>);
decode(pad, <<A0:8, B0:8, C0:8, $=:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0) bsr 2,
  decode(pad, <<>>, <<Acc/binary, A:6, B:6, C:4>>);
decode(Mode, <<A0:8, B0:8, C0:8, D0:8, Rest/binary>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0),
  D = dec_base64_char(D0),
  decode(Mode, Rest, <<Acc/binary, A:6, B:6, C:6, D:6>>);
decode(_Mode, Data, _Acc) ->
  {error, {invalid_data, Data}}.

-spec dec_base64_char($A..$Z | $a..$z | $0..$9 | $+ | $/) ->
        0..63.
dec_base64_char(Char) when Char >= $A, Char =< $Z ->
  Char - $A;
dec_base64_char(Char) when Char >= $a, Char =< $z ->
  Char - $a + 26;
dec_base64_char(Char) when Char >= $0, Char =< $9 ->
  Char - $0 + 52;
dec_base64_char($+) ->
  62;
dec_base64_char($/) ->
  63;
dec_base64_char(Char) ->
  throw({error, {invalid_base64_char, Char}}).


src/b64url.erl
%% Copyright (c) 2022 Bryan Frimin <bryan@frimin.fr>.
%% Copyright (c) 2021 Exograd SAS.
%% Copyright (c) 2021 Bryan Frimin <bryan@frimin.fr>.
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
%% SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR
%% IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(b64url).

-export([encode/1, encode/2,
         decode/1, decode/2]).

-export_type([decode_error_reason/0]).

-type decode_error_reason() ::
        invalid_encoding
      | {invalid_data, binary()}
      | {invalid_base64_char, byte()}.

-spec encode(binary()) -> binary().
encode(Bin) when is_binary(Bin) ->
  encode(Bin, []).

-spec encode(binary(), Options) -> binary() when
    Options :: [Option],
    Option :: nopad.
encode(Bin, Options) when is_binary(Bin), is_list(Options) ->
  case proplists:get_bool(nopad, Options) of
    true ->
      encode(nopad, Bin, <<>>);
    false ->
      encode(pad, Bin, <<>>)
  end.

-spec encode(pad | nopad, binary(), binary()) -> binary().
encode(_Mode, <<>>, Acc) ->
  Acc;
encode(Mode, <<A0:6, B0:6, C0:6, D0:6, Rest/binary>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0),
  D = dec_base64_digit(D0),
  encode(Mode, Rest, <<Acc/binary, A, B, C, D>>);
encode(nopad, <<A0:6, B0:2>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0 bsl 4),
  encode(nopad, <<>>, <<Acc/binary, A, B>>);
encode(nopad, <<A0:6, B0:6, C0:4>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0 bsl 2),
  encode(nopad, <<>>, <<Acc/binary, A, B, C>>);
encode(pad, <<A0:6, B0:2>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0 bsl 4),
  encode(pad, <<>>, <<Acc/binary, A, B, $=, $=>>);
encode(pad, <<A0:6, B0:6, C0:4>>, Acc) ->
  A = dec_base64_digit(A0),
  B = dec_base64_digit(B0),
  C = dec_base64_digit(C0 bsl 2),
  encode(pad, <<>>, <<Acc/binary, A, B, C, $=>>).

-spec dec_base64_digit(0..63) ->
        $A..$Z | $a..$z | $0..$9 | $- | $_.
dec_base64_digit(Char) when Char =< 25 ->
  Char + $A;
dec_base64_digit(Char) when Char =< 51 ->
  Char + $a - 26;
dec_base64_digit(Char) when Char =< 61 ->
  Char + $0 - 52;
dec_base64_digit(62) ->
  $-;
dec_base64_digit(63) ->
  $_.

-spec decode(binary()) ->
        {ok, binary()} | {error, decode_error_reason()}.
decode(Bin) when is_binary(Bin) ->
  decode(Bin, []).

-spec decode(binary(), Options) ->
        {ok, binary()} | {error, decode_error_reason()}
          when Options :: [Option],
               Option :: nopad.
decode(Bin, Options) when is_binary(Bin), is_list(Options) ->
  try
    case proplists:get_bool(nopad, Options) of
      true ->
        decode(nopad, Bin, <<>>);
      false ->
        decode(pad, Bin, <<>>)
    end
  catch
    throw:{error, Reason} ->
      {error, Reason}
  end.

-spec decode(pad | nopad, binary(), binary()) ->
        {ok, binary()} | {error, term()}.
decode(_Mode, <<>>, Acc) ->
  {ok, Acc};
decode(nopad, <<_A:8, _B:8, _C:8, $=:8>>, _Acc) ->
  {error, invalid_encoding};
decode(nopad, <<_A:8, _B:8, $=:8, $=:8>>, _Acc) ->
  {error, invalid_encoding};
decode(nopad, <<A0:8, B0:8, C0:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0) bsr 2,
  decode(nopad, <<>>, <<Acc/binary, A:6, B:6, C:4>>);
decode(nopad, <<A0:8, B0:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0) bsr 4,
  decode(nopad, <<>>, <<Acc/binary, A:6, B:2>>);
decode(pad, <<A0:8, B0:8, $=:8, $=:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0) bsr 4,
  decode(pad, <<>>, <<Acc/binary, A:6, B:2>>);
decode(pad, <<A0:8, B0:8, C0:8, $=:8>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0) bsr 2,
  decode(pad, <<>>, <<Acc/binary, A:6, B:6, C:4>>);
decode(Mode, <<A0:8, B0:8, C0:8, D0:8, Rest/binary>>, Acc) ->
  A = dec_base64_char(A0),
  B = dec_base64_char(B0),
  C = dec_base64_char(C0),
  D = dec_base64_char(D0),
  decode(Mode, Rest, <<Acc/binary, A:6, B:6, C:6, D:6>>);
decode(_Mode, Data, _Acc) ->
  {error, {invalid_data, Data}}.

-spec dec_base64_char($A..$Z | $a..$z | $0..$9 | $- | $_) ->
        0..63.
dec_base64_char(Char) when Char >= $A, Char =< $Z ->
  Char - $A;
dec_base64_char(Char) when Char >= $a, Char =< $z ->
  Char - $a + 26;
dec_base64_char(Char) when Char >= $0, Char =< $9 ->
  Char - $0 + 52;
dec_base64_char($-) ->
  62;
dec_base64_char($_) ->
  63;
dec_base64_char(Char) ->
  throw({error, {invalid_base64_char, Char}}).
